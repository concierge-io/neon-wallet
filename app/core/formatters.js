// @flow
import { truncateNumber } from './math'

const GAS_DECIMAL_LENGTH = 7
const GAS_DECIMAL_SHORT_DISPLAY_LENGTH = 4

const FIAT_DECIMAL_LENGTH = 2

export const formatGAS = (gas: number | string, shortDisplay: boolean = false): string => {
  const decimalLength = shortDisplay ? GAS_DECIMAL_SHORT_DISPLAY_LENGTH : GAS_DECIMAL_LENGTH
  return truncateNumber(parseFloat(gas), decimalLength).toFixed(decimalLength)
}

export const formatNEO = (neo: number | string): number => parseInt(neo, 10)

export const formatFiat = (value: number | string): string => parseFloat(value).toFixed(FIAT_DECIMAL_LENGTH)
