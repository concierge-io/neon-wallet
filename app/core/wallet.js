// @flow
import { ASSETS_LABELS } from './constants'
import { getAccountFromWIFKey, verifyAddress } from 'neon-js'

const MIN_PASSPHRASE_LEN = 4

export const validatePassphrase = (passphrase: string): boolean => passphrase.length >= MIN_PASSPHRASE_LEN

export const checkMatchingPassphrases = (passphrase: string, passphrase2: string) => passphrase !== passphrase2

export const verifyPrivateKey = (wif: string): boolean => {
  if (!wif) {
    return false
  }
  const account = getAccountFromWIFKey(wif)
  return account !== -1 && account.address
}

export const validateTransactionBeforeSending = (neoBalance: number, gasBalance: number, selectedAsset: string, selectedHash: string, balances: Object[], sendAddress: string, sendAmount: string) => {
  console.log('entered validation');
  if (!sendAddress || !sendAmount) {
    return {
      error: 'Please specify an address and amount',
      valid: false
    }
  }

  if (verifyAddress(sendAddress) !== true || sendAddress.charAt(0) !== 'A') {
    return {
      error: 'The address you entered was not valid.',
      valid: false
    }
  }

  if (selectedAsset === ASSETS_LABELS.NEO) {
    if (parseFloat(sendAmount) !== parseInt(sendAmount)) { // check for fractional neo
      return {
        error: 'You cannot send fractional amounts of Neo.',
        valid: false
      }
    }
    if (parseInt(sendAmount) > neoBalance) { // check for value greater than account balance
      return {
        error: 'You do not have enough NEO to send.',
        valid: false
      }
    }
  } else if (selectedAsset === ASSETS_LABELS.GAS) {
    if (parseFloat(sendAmount) > gasBalance) {
      return {
        error: 'You do not have enough GAS to send.',
        valid: false
      }
    }
  } else if(selectedHash){
    console.log('these are the balances', balances)
    if ( !balances[selectedHash] ||  parseFloat(sendAmount) > parseFloat(balances[selectedHash]) ){
      return {
        error: 'You do not have enough to send.',
        valid: false
      }
    }
  }

  if (parseFloat(sendAmount) < 0) { // check for negative asset
    return {
      error: 'You cannot send negative amounts of an asset.',
      valid: false
    }
  }

  return {
    error: '',
    valid: true
  }
}
