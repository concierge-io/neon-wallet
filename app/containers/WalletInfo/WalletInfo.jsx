// @flow
import React, { Component } from 'react'
import { isNil } from 'lodash'
import Claim from '../Claim'
import MdSync from 'react-icons/lib/md/sync'
import QRCode from 'qrcode/lib/browser'
import { clipboard } from 'electron'
import Copy from 'react-icons/lib/md/content-copy'
import ReactTooltip from 'react-tooltip'
import { formatGAS, formatFiat } from '../../core/formatters'
import { ASSETS } from '../../core/constants'

type Props = {
  address: string,
  neo: number,
  net: NetworkType,
  gas: number,
  neoPrice: number,
  gasPrice: number,
  initiateGetBalance: Function,
  showInfoNotification: Function,
  showSuccessNotification: Function,
  showErrorNotification: Function
}

export default class WalletInfo extends Component<Props> {
  canvas: ?HTMLCanvasElement

  componentDidMount () {
    const { initiateGetBalance, net, address } = this.props
    initiateGetBalance(net, address)
    QRCode.toCanvas(this.canvas, address, { version: 5 }, (err) => {
      if (err) console.log(err)
    })
  }

  // force sync with balance data
  refreshBalance = () => {
    const {
      showInfoNotification,
      showSuccessNotification,
      showErrorNotification,
      initiateGetBalance,
      net,
      address
    } = this.props
    showInfoNotification({ message: 'Retrieving blockchain information...' })
    initiateGetBalance(net, address).then((response) => {
      showSuccessNotification({ message: 'Received latest blockchain information.' })
    }).catch(() => {
      showErrorNotification({ message: 'Failed to retrieve blockchain information' })
    })
  }

  render () {
    const { address, neo, gas, neoPrice, gasPrice } = this.props
    if (isNil(address)) {
      return null
    }

    let neoValue = 0
    let gasValue = 0
    let totalValue = 0

    if (neoPrice && neo) {
      neoValue = formatFiat(neoPrice * neo)
      if (gasPrice && gas) {
        gasValue = formatFiat(gasPrice * gas)
        totalValue = formatFiat(neoPrice * neo + gasPrice * gas)
      }
    }

    return (
      <div id='accountInfo'>
        <div className='label'>Your Public Neo Address:</div>
        <div className='address'>
          {address}
          <span className='copyKey' onClick={() => clipboard.writeText(address)}><Copy data-tip data-for='copyAddressTip' /></span>
        </div>
        <ReactTooltip class='solidTip' id='copyAddressTip' place='bottom' type='dark' effect='solid'>
          <span>Copy Public Address</span>
        </ReactTooltip>
        <div className='spacer' />
        <div id='balance'>
          <div className='split'>
            <div className='label'>{ASSETS.NEO}</div>
            <div className='amountBig amountNeo'>{neo}</div>
            <div className='fiat neoWalletValue'>US ${neoValue}</div>
          </div>
          <div className='split'>
            <div className='label'>{ASSETS.GAS}</div>
            <div className='amountBig amountGas' data-tip data-for='gasBalanceTip'>{formatGAS(gas, true)}</div>
            <ReactTooltip class='solidTip' id='gasBalanceTip' place='bottom' type='dark' effect='solid' disable={gas === 0}>
              <span className='amountGasTooltip'>{formatGAS(gas)}</span>
            </ReactTooltip>
            <div className='fiat gasWalletValue'>US ${gasValue}</div>
          </div>
          <div className='fiat walletTotal'>Total US ${totalValue}</div>
          <div className='refreshBalance' onClick={this.refreshBalance} >
            <MdSync id='refresh' data-tip data-for='refreshBalanceTip' />
            <ReactTooltip class='solidTip' id='refreshBalanceTip' place='bottom' type='dark' effect='solid'>
              <span>Refresh account balance</span>
            </ReactTooltip>
          </div>
        </div>
        <div className='spacer' />
        <Claim />
        <div className='spacer' />
        <div className='qrCode'><canvas id='qrCanvas' ref={(node) => { this.canvas = node }} /></div>
      </div>
    )
  }
}
