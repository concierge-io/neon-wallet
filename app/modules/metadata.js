// @flow
import { getWalletDBHeight, getAPIEndpoint } from 'neon-js'
import axios from 'axios'
import { version } from '../../package.json'
import { showWarningNotification } from './notifications'
import { NETWORK, EXPLORER, NEON_WALLET_RELEASE_LINK, NOTIFICATION_POSITIONS } from '../core/constants'
import asyncWrap from '../core/asyncHelper'

// Constants
export const SET_HEIGHT = 'SET_HEIGHT'
export const SET_NETWORK = 'SET_NETWORK'
export const SET_EXPLORER = 'SET_EXPLORER'

// Actions
export function setNetwork (net: NetworkType) {
  const network = net === NETWORK.MAIN? NETWORK.MAIN : net === NETWORK.APHELION_MAIN? NETWORK.APHELION_MAIN: net === NETWORK.APHELION_TEST? NETWORK.APHELION_TEST: NETWORK.TEST
  return {
    type: SET_NETWORK,
    payload: { network }
  }
}

export function setBlockHeight (blockHeight: number) {
  return {
    type: SET_HEIGHT,
    payload: { blockHeight }
  }
}

export function setBlockExplorer (blockExplorer: ExplorerType) {
  return {
    type: SET_EXPLORER,
    payload: { blockExplorer }
  }
}

export const checkVersion = () => async (dispatch: DispatchType, getState: GetStateType) => {
  const state = getState().metadata
  const { net } = state
  const apiEndpoint = getAPIEndpoint(net)

  const [err, res] = await asyncWrap(axios.get(`${apiEndpoint}/v2/version`))
  const shouldUpdate = res && res.data && res.data.version !== version
  if (err || shouldUpdate) {
    const link = `<a href='${NEON_WALLET_RELEASE_LINK}' target='_blank' class="notification-link">${NEON_WALLET_RELEASE_LINK}</a>`
    const message = err ? `Error checking wallet version! Please make sure you have downloaded the latest version: ${link}`
      : `Your wallet is out of date! Please download the latest version from ${link}`
    return dispatch(showWarningNotification({
      message,
      autoDismiss: 0,
      stack: true,
      position: NOTIFICATION_POSITIONS.BOTTOM_CENTER
    }))
  }
}

export const syncBlockHeight = (net: NetworkType) => async (dispatch: DispatchType) => {
  const [_err, blockHeight] = await asyncWrap(getWalletDBHeight(net)) // eslint-disable-line
  return dispatch(setBlockHeight(blockHeight))
}

// state getters
export const getBlockHeight = (state) => state.metadata.blockHeight
export const getNetwork = (state) => state.metadata.network
export const getBlockExplorer = (state) => state.metadata.blockExplorer

const initialState = {
  blockHeight: 0,
  network: NETWORK.MAIN,
  blockExplorer: EXPLORER.NEO_TRACKER
}

export default (state: Object = initialState, action: Object) => {
  switch (action.type) {
    case SET_HEIGHT:
      const { blockHeight } = action.payload
      return {
        ...state,
        blockHeight
      }
    case SET_EXPLORER:
      const { blockExplorer } = action.payload
      return {
        ...state,
        blockExplorer
      }
    case SET_NETWORK:
      const { network } = action.payload
      return {
        ...state,
        network
      }
    default:
      return state
  }
}
