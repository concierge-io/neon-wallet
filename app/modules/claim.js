// @flow
import { doClaimAllGas, doSendAsset, getClaimAmounts, hardwareDoSendAsset, hardwareDoClaimAllGas } from 'neon-js'
import { log } from '../util/Logs'
import { ASSETS } from '../core/constants'
import { showErrorNotification, showSuccessNotification, showInfoNotification } from './notifications'
import { FIVE_MINUTES_MS } from '../core/time'
import { getWif, getAddress, getSigningFunction, getPublicKey, LOGOUT } from './account'
import { getNetwork } from './metadata'
import { getNeo } from './wallet'
import asyncWrap from '../core/asyncHelper'

// Constants
export const SET_CLAIM = 'SET_CLAIM'
export const SET_CLAIM_REQUEST = 'SET_CLAIM_REQUEST'
export const DISABLE_CLAIM = 'DISABLE_CLAIM'

// Actions
export function setClaim (claimAvailable: number, claimUnavailable: number) {
  return {
    type: SET_CLAIM,
    payload: {
      claimAvailable,
      claimUnavailable
    }
  }
}

export function setClaimRequest (claimRequest: boolean) {
  return {
    type: SET_CLAIM_REQUEST,
    payload: { claimRequest }
  }
}

export function disableClaim (disableClaimButton: boolean) {
  return {
    type: DISABLE_CLAIM,
    payload: { disableClaimButton }
  }
}

export const syncAvailableClaim = (net: NetworkType, address: string) => async (dispatch: DispatchType) => {
  const [_err, result] = await asyncWrap(getClaimAmounts(net, address)) // eslint-disable-line
  return dispatch(setClaim(result.available, result.unavailable))
}

export const doClaimNotify = () => async (dispatch: DispatchType, getState: GetStateType) => {
  const state = getState()
  const wif = getWif(state)
  const address = getAddress(state)
  const net = getNetwork(state)
  const signingFunction = getSigningFunction(state)
  const publicKey = getPublicKey(state)

  log(net, 'CLAIM', address, { info: 'claim all gas' })

  const isHardwareClaim = !!publicKey

  let claimGasFn
  if (isHardwareClaim) {
    dispatch(showInfoNotification({
      message: 'Sign transaction 2 of 2 to claim Gas on your hardware device (claiming Gas)',
      autoDismiss: 0
    }))
    claimGasFn = () => hardwareDoClaimAllGas(net, publicKey, signingFunction)
  } else {
    claimGasFn = () => doClaimAllGas(net, wif)
  }

  const [err, response] = await asyncWrap(claimGasFn())
  if (!err && response.result) {
    dispatch(showSuccessNotification({
      message: 'Claim was successful! Your balance will update once the blockchain has processed it.'
    }))
    setTimeout(() => dispatch(disableClaim(false)), FIVE_MINUTES_MS)
  } else {
    return dispatch(showErrorNotification({ message: 'Claim failed' }))
  }
}

// To initiate claim, first send all Neo to own address, the set claimRequest state
// When new claims are available, this will trigger the claim
export const doGasClaim = () => async (dispatch: DispatchType, getState: GetStateType) => {
  const state = getState()
  const wif = getWif(state)
  const address = getAddress(state)
  const net = getNetwork(state)
  const neo = getNeo(state)
  const signingFunction = getSigningFunction(state)
  const publicKey = getPublicKey(state)

  // if no neo in account, no need to send to self first
  if (neo === 0) {
    return dispatch(doClaimNotify())
  } else {
    dispatch(showInfoNotification({ message: 'Sending Neo to Yourself...', autoDismiss: 0 }))
    log(net, 'SEND', address, { to: address, amount: neo, asset: ASSETS.NEO })

    const isHardwareClaim = !!publicKey

    let sendAssetFn
    if (isHardwareClaim) {
      dispatch(showInfoNotification({
        message: 'Sign transaction 1 of 2 to claim Gas on your hardware device (sending Neo to yourself)',
        autoDismiss: 0
      }))
      sendAssetFn = () => hardwareDoSendAsset(net, address, publicKey, { [ASSETS.NEO]: neo }, signingFunction)
    } else {
      sendAssetFn = () => doSendAsset(net, address, wif, { [ASSETS.NEO]: neo })
    }

    const [err, response] = await asyncWrap(sendAssetFn())
    if (err || response.result === undefined || response.result === false) {
      return dispatch(showErrorNotification({ message: 'Transaction failed!' }))
    } else {
      dispatch(showInfoNotification({ message: 'Waiting for transaction to clear...', autoDismiss: 0 }))
      dispatch(setClaimRequest(true))
      return dispatch(disableClaim(true))
    }
  }
}

// State Getters
export const getClaimRequest = (state) => state.claim.claimRequest
export const getClaimAmount = (state) => state.claim.claimAmount
export const getClaimAvailable = (state) => state.claim.claimAvailable
export const getClaimUnavailable = (state) => state.claim.claimUnavailable
export const getClaimWasUpdated = (state) => state.claim.claimWasUpdated
export const getDisableClaimButton = (state) => state.claim.disableClaimButton

const initialState = {
  claimRequest: false,
  claimAmount: 0,
  claimAvailable: 0,
  claimUnavailable: 0,
  claimWasUpdated: false,
  disableClaimButton: false
}

export default (state: Object = initialState, action: Object) => {
  switch (action.type) {
    case SET_CLAIM_REQUEST:
      const { claimRequest } = action.payload
      return {
        ...state,
        claimRequest
      }
    case SET_CLAIM:
      const { claimAvailable, claimUnavailable } = action.payload
      let claimWasUpdated = false
      if (claimAvailable > state.claimAvailable && state.claimRequest === true) {
        claimWasUpdated = true
      }
      return {
        ...state,
        claimAmount: (claimAvailable + claimUnavailable) / 100000000,
        claimAvailable,
        claimUnavailable,
        claimWasUpdated
      }
    case DISABLE_CLAIM:
      const { disableClaimButton } = action.payload
      return {
        ...state,
        disableClaimButton
      }
    case LOGOUT:
      return initialState
    default:
      return state
  }
}
