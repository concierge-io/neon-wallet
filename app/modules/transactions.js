// @flow
/* eslint-disable camelcase */

import { ASSETS_LABELS, ASSETS, ASSET_ID } from '../core/constants'
import { validateTransactionBeforeSending } from '../core/wallet'
import { getTransactionHistory, doSendAsset, hardwareDoSendAsset, getRPCEndpoint, getBalance, serializeTransaction } from 'neon-js'
import {setTransactionHistory, getNeo, getGas, createCoin} from './wallet'
import { log } from '../util/Logs'
import { showErrorNotification, showInfoNotification, showSuccessNotification } from './notifications'
import { getWif, getPublicKey, getSigningFunction, getAddress, LOGOUT } from './account'
import { getNetwork } from './metadata'
import asyncWrap from '../core/asyncHelper'
import { buildScript } from '../sc/scriptBuilder'
import { reverseHex } from 'neon-js/src/utils'
import { doInvokeScript } from 'neon-js/src/api'
import { getScriptHashFromAddress, getAccountFromWIFKey } from 'neon-js/src/wallet'
import { invocationTx } from 'neon-js/src/transactions/create'
import { signTransaction } from 'neon-js/src/transactions/index'
import { Query } from '../core/query'

// Constants
export const TOGGLE_ASSET = 'TOGGLE_ASSET'
export const LOADING_TRANSACTIONS = 'LOADING_TRANSACTIONS'

export const toggleAsset = ( symbol: string,  hash: string) => ({
  type: TOGGLE_ASSET,
  asset: symbol,
  hash: hash
})

export const setIsLoadingTransaction = (isLoading: boolean) => ({
  type: LOADING_TRANSACTIONS,
  payload: {
    isLoadingTransactions: isLoading
  }
})

export const syncTransactionHistory = (net: NetworkType, address: string) => async (dispatch: DispatchType) => {
  dispatch(setIsLoadingTransaction(true))
  const [err, transactions] = await asyncWrap(getTransactionHistory(net, address))
  if (!err && transactions) {
    const txs = transactions.map(({ NEO, GAS, txid, block_index, neo_sent, neo_gas }: TransactionHistoryType) => ({
      type: neo_sent ? ASSETS.NEO : ASSETS.GAS,
      amount: neo_sent ? NEO : GAS,
      txid,
      block_index
    }))
    dispatch(setIsLoadingTransaction(false))
    dispatch(setTransactionHistory(txs))
  } else {
    dispatch(setIsLoadingTransaction(false))
  }
}

export const sendTransaction = (sendAddress: string, sendAmount: string) => async (dispatch: DispatchType, getState: GetStateType): Promise<*> => {

  const state = getState()
  const wif = getWif(state)
  const address = getAddress(state)
  const net = getNetwork(state)
  const neo = getNeo(state)
  const gas = getGas(state)
  const selectedAsset = getSelectedAsset(state)
  const selectedHash = getSelectedHash(state)
  const balances = state.nep.balances
  const signingFunction = getSigningFunction(state)
  const account = getAccountFromWIFKey(wif)
  const publicKey = account.publicKeyEncoded
  const privateKey = account.privateKey
  const tokens = state.nep.tokens

  const rejectTransaction = (message: string) => dispatch(showErrorNotification({ message }))
  const { error, valid } = validateTransactionBeforeSending(neo, gas, selectedAsset, selectedHash, balances, sendAddress, sendAmount)

  if (valid) {
    if(selectedHash){

      let parsedValue = Math.round(sendAmount * Math.pow(10, tokens[selectedHash].decimals))

      //run this logic for nep5 contracts
      doTransferToken(net, selectedHash.slice(2), wif, publicKey, privateKey, address, sendAddress, parsedValue, 0, signingFunction)
        .then((result) => {

          if(result.result){
            return dispatch(showSuccessNotification({ message: 'The transaction was successful. Wait until the transfer is approved to see it on the balance.' }))
          }else{
            return rejectTransaction('Transaction failed!')
          }
      })

    }else{

      const selfAddress = address
      const assetName = selectedAsset === ASSETS_LABELS.NEO ? ASSETS.NEO : ASSETS.GAS
      let sendAsset = {}
      sendAsset[assetName] = sendAmount

      dispatch(showInfoNotification({ message: 'Sending Transaction...', autoDismiss: 0 }))
      log(net, 'SEND', selfAddress, { to: sendAddress, asset: selectedAsset, amount: sendAmount })

      const isHardwareSend = !publicKey

      let sendAssetFn
      if (isHardwareSend) {
        dispatch(showInfoNotification({ message: 'Please sign the transaction on your hardware device', autoDismiss: 0 }))
        sendAssetFn = () => hardwareDoSendAsset(net, sendAddress, publicKey, sendAsset, signingFunction)
      } else {
        sendAssetFn = () => doSendAsset(net, sendAddress, wif, sendAsset)
      }

      const [err, response] = await asyncWrap(sendAssetFn())
      if (err || response.result === undefined || response.result === false) {
        return rejectTransaction('Transaction failed!')
      } else {
        return dispatch(showSuccessNotification({ message: 'Transaction complete! Your balance will automatically update when the blockchain has processed it.' }))
      }
    }

  } else {
    return rejectTransaction(error)
  }
}

/**
 * Transfers NEP5 Tokens.
 * @param {string} net
 * @param {string} scriptHash
 * @param {string} wif
 * @param {string} publickey
 * @param {string} fromAddress
 * @param {string} toAddress
 * @param {number} transferAmount
 * @param {number} gasCost
 * @param {function} signingFunction
 * @return {Promise<Response>} RPC response
 */
const doTransferToken = (net, scriptHash, wif, publickey, privatekey, fromAddress, toAddress, transferAmount, gasCost = 0, signingFunction = null) => {
  const rpcEndpointPromise = getRPCEndpoint(net)
  const balancePromise = getBalance(net, fromAddress)
  let signedTx
  let endpt
  return Promise.all([rpcEndpointPromise, balancePromise])
    .then((values) => {
      endpt = values[0]
      const balances = values[1]
      const fromAddrScriptHash = reverseHex(getScriptHashFromAddress(fromAddress))
      const intents = [
        { assetId: ASSET_ID.GAS, value: 0.00000001, scriptHash: fromAddrScriptHash }
      ]
      const toAddrScriptHash = reverseHex(getScriptHashFromAddress(toAddress))
      const invoke = { scriptHash, operation: 'transfer', args: [fromAddrScriptHash, toAddrScriptHash, transferAmount] }
      const unsignedTx = invocationTx(publickey, balances, intents, invoke, gasCost, { version: 1 })
      if (signingFunction) {
        return signingFunction(unsignedTx, publickey)
      } else {
        return signTransaction(unsignedTx, privatekey)
      }
    })
    .then((signedResult) => {
      signedTx = signedResult
      return sendRawTransaction(signedTx).execute(endpt)
    })
    .then((res) => {
      if (res.result === true) {
        res.txid = signedTx
      }
      return res
    })
}

export const sendRawTransaction = (transaction) => {
  const serialized = typeof (transaction) === 'object' ? serializeTransaction(transaction) : transaction
  return new Query({
    method: 'sendrawtransaction',
    params: [serialized]
  })
}

// state getters
export const getSelectedAsset = (state) => state.transactions.selectedAsset
export const getSelectedHash = (state) => state.transactions.selectedHash
export const getIsLoadingTransactions = (state) => state.transactions.isLoadingTransactions

const initialState = {
  selectedAsset: ASSETS_LABELS.NEO,
  selectedHash: null,
  isLoadingTransactions: false
}

export default (state: Object = initialState, action: Object) => {
  switch (action.type) {
    case LOADING_TRANSACTIONS:
      const { isLoadingTransactions } = action.payload
      return {
        ...state,
        isLoadingTransactions
      }
    case TOGGLE_ASSET:
      return {
        ...state,
        selectedAsset: action.asset,
        selectedHash: action.hash
      }
    case LOGOUT:
      return initialState
    default:
      return state
  }
}
