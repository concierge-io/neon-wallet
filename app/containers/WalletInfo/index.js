// @flow
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import WalletInfo from './WalletInfo'
import { initiateGetBalance, getNeo, getGas, getNeoPrice, getGasPrice } from '../../modules/wallet'
import { showErrorNotification, showSuccessNotification, showInfoNotification } from '../../modules/notifications'
import { getAddress } from '../../modules/account'
import { getNetwork } from '../../modules/metadata'

const mapStateToProps = (state) => ({
  neo: getNeo(state),
  gas: getGas(state),
  address: getAddress(state),
  net: getNetwork(state),
  neoPrice: getNeoPrice(state),
  gasPrice: getGasPrice(state)
})

const actionCreators = {
  initiateGetBalance,
  showErrorNotification,
  showSuccessNotification,
  showInfoNotification
}

const mapDispatchToProps = dispatch => bindActionCreators(actionCreators, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(WalletInfo)
