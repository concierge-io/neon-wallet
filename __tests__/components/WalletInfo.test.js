import React from 'react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { mount, shallow } from 'enzyme'
import { SET_TRANSACTION_HISTORY, SET_BALANCE, SET_GAS_PRICE, SET_NEO_PRICE } from '../../app/modules/wallet'
import { SHOW_NOTIFICATION, HIDE_NOTIFICATIONS } from '../../app/modules/notifications'
import { LOADING_TRANSACTIONS } from '../../app/modules/transactions'
import { SET_HEIGHT } from '../../app/modules/metadata'
import { SET_CLAIM } from '../../app/modules/claim'
import WalletInfo from '../../app/containers/WalletInfo'

// TODO research how to move the axios mock code which is repeated in NetworkSwitch to a helper or config file
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { clipboard } from 'electron'
import { version } from '../../package.json'
import { formatFiat } from '../../app/core/formatters'

const axiosMock = new MockAdapter(axios)
axiosMock
  .onGet('http://testnet-api.wallet.cityofzion.io/v2/version')
  .reply(200, { version })
axiosMock
  .onGet('https://api.coinmarketcap.com/v1/ticker/neo/?convert=USD')
  .reply(200, [ { price_usd: 24.50 } ])
axiosMock
  .onGet('https://api.coinmarketcap.com/v1/ticker/gas/?convert=USD')
  .reply(200, [ { price_usd: 18.20 } ])

jest.mock('electron', () => ({
  clipboard: {
    writeText: jest.fn()
  }
}))
jest.useFakeTimers()

jest.unmock('qrcode')
import QRCode from 'qrcode/lib/browser' // eslint-disable-line
QRCode.toCanvas = jest.fn()

const initialState = {
  account: {
    address: 'ANqUrhv99rwCiFTL6N1An9NH5UVkPYxTuw'
  },
  metadata: {
    network: 'TestNet'
  },
  wallet: {
    Neo: 10,
    Gas: 1.0001001,
    neoPrice: 25.48,
    gasPrice: 18.10
  },
  claim: {
    claimAmount: 0.5
  }
}

const setup = (state = initialState, shallowRender = true) => {
  const store = configureStore([thunk])(state)

  let wrapper
  if (shallowRender) {
    wrapper = shallow(<WalletInfo store={store} />)
  } else {
    wrapper = mount(
      <Provider store={store}>
        <WalletInfo />
      </Provider>
    )
  }

  return {
    store,
    wrapper
  }
}

describe('WalletInfo', () => {
  test('renders without crashing', (done) => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
    done()
  })
  test('correctly renders data from state', (done) => {
    const { wrapper } = setup(initialState, false)

    const addressField = wrapper.find('.address')
    const neoWalletValue = wrapper.find('.neoWalletValue')
    const gasWalletValue = wrapper.find('.gasWalletValue')
    const walletValue = wrapper.find('.walletTotal')
    const expectedNeoWalletValue = formatFiat(initialState.wallet.neoPrice * initialState.wallet.Neo)
    const expectedGasWalletValue = formatFiat(initialState.wallet.gasPrice * initialState.wallet.Gas)
    const expectedWalletValue = formatFiat(initialState.wallet.neoPrice * initialState.wallet.Neo + initialState.wallet.gasPrice * initialState.wallet.Gas)
    const neoField = wrapper.find('.amountNeo')
    const gasField = wrapper.find('.amountGas')

    expect(neoWalletValue.text()).toEqual(`US $${expectedNeoWalletValue}`)
    expect(gasWalletValue.text()).toEqual(`US $${expectedGasWalletValue}`)
    expect(walletValue.text()).toEqual(`Total US $${expectedWalletValue}`)
    expect(addressField.text().split('<')[0]).toEqual(initialState.account.address)
    expect(neoField.text()).toEqual(`${initialState.wallet.Neo}`)
    // TODO: Test the gas tooltip value, this is testing the display value, truncated to 4 decimals
    expect(gasField.text()).toEqual('1.0001')
    done()
  })
  test('copy to clipboard is getting called on click', async () => {
    const { wrapper } = setup()
    const deepWrapper = wrapper.dive()

    expect(clipboard.writeText.mock.calls.length).toBe(0)
    deepWrapper.find('.copyKey').simulate('click')
    await Promise.resolve('Pause').then()
    expect(clipboard.writeText.mock.calls.length).toBe(1)
  })
  test('refreshBalance is getting called on click', async () => {
    const { wrapper, store } = setup()
    const deepWrapper = wrapper.dive()

    const actionTypes = [
      HIDE_NOTIFICATIONS,
      SHOW_NOTIFICATION,
      LOADING_TRANSACTIONS,
      SET_TRANSACTION_HISTORY,
      SET_HEIGHT,
      SET_NEO_PRICE,
      SET_GAS_PRICE,
      SET_BALANCE,
      SET_CLAIM
    ]
    deepWrapper.find('.refreshBalance').simulate('click')
    jest.runAllTimers()
    await Promise.resolve('Pause').then().then().then().then()
    const actions = store.getActions()
    expect(actions.length).toEqual(16)
    // expect(actions.length).toEqual(12)
    actions.forEach(action => {
      expect(actionTypes.indexOf(action.type) > -1).toEqual(true)
    })
  })
  test('calls the correct number of actions after mounting', async () => {
    const { store } = setup(initialState, false)
    const actionTypes = [
      SET_TRANSACTION_HISTORY,
      SET_HEIGHT,
      SET_CLAIM,
      SET_BALANCE,
      LOADING_TRANSACTIONS
    ]

    jest.runAllTimers()
    await Promise.resolve('Pause').then().then().then()
    const actions = store.getActions()
    expect(actions.length).toEqual(6)
    actions.forEach(action => {
      expect(actionTypes.indexOf(action.type) > -1).toEqual(true)
    })
  })
})
