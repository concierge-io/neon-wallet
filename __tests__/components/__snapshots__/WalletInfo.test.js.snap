// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletInfo renders without crashing 1`] = `
<WalletInfo
  address="ANqUrhv99rwCiFTL6N1An9NH5UVkPYxTuw"
  gas={1.0001001}
  gasPrice={18.1}
  initiateGetBalance={[Function]}
  neo={10}
  neoPrice={25.48}
  net="TestNet"
  showErrorNotification={[Function]}
  showInfoNotification={[Function]}
  showSuccessNotification={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
